File: Code_CS/C14_Personalization/DisplayModeMenu.ascx.cs
using System;
using System.Web.UI;
using System.Web.UI.WebControls;
using System.Web.UI.WebControls.WebParts;

public partial class DisplayModeMenu : UserControl
{

   // will reference the current WebPartManager control.    
   WebPartManager webPartManager;

   public void Page_Init(object sender, EventArgs e)
   {
      Page.InitComplete += new EventHandler(InitComplete);
   }

   // when the page is fully initialized    
   public void InitComplete(object sender, EventArgs e)
   {
      webPartManager = WebPartManager.GetCurrentWebPartManager(Page);
      String browseModeName = WebPartManager.BrowseDisplayMode.Name;
      foreach (WebPartDisplayMode mode in webPartManager.SupportedDisplayModes)
      {
         String modeName = mode.Name;
         if (mode.IsEnabled(webPartManager))
         {
            ListItem listItem = new ListItem(modeName, modeName);
            ddlDisplayMode.Items.Add(listItem);
         }
      }
   }

   // Change the page to the selected display mode.    
   public void ddlDisplayMode_SelectedIndexChanged(object sender, EventArgs e)
   {
      String selectedMode = ddlDisplayMode.SelectedValue;
      WebPartDisplayMode mode = webPartManager.SupportedDisplayModes[selectedMode];
      if (mode != null)
      {
         webPartManager.DisplayMode = mode;
      }
   }

   // Set the selected item equal to the current display mode.   
   public void Page_PreRender(object sender, EventArgs e)
   {
      ListItemCollection items = ddlDisplayMode.Items;
      int selectedIndex = items.IndexOf(items.FindByText(webPartManager.DisplayMode.Name));
      ddlDisplayMode.SelectedIndex = selectedIndex;
   }
}

File: Code_VB/C14_Personalization/DisplayModeMenu.ascx.vb

Partial Class DisplayModeMenu
    Inherits System.Web.UI.UserControl

    ' will reference the current WebPartManager control. 
    Private webPartManager As WebPartManager


    Protected Sub Page_Init1(ByVal sender As Object, ByVal e As System.EventArgs) Handles Me.Init

        AddHandler Page.InitComplete, AddressOf InitComplete

    End Sub

    ' when the page is fully initialized 
    Public Sub InitComplete(ByVal sender As Object, ByVal e As EventArgs)

        webPartManager = webPartManager.GetCurrentWebPartManager(Page)
        Dim browseModeName As String = webPartManager.BrowseDisplayMode.Name

        For Each mode As WebPartDisplayMode In webPartManager.SupportedDisplayModes
            Dim modeName As String = mode.Name
            If mode.IsEnabled(webPartManager) Then
                Dim listItem As New ListItem(modeName, modeName)
                ddlDisplayMode.Items.Add(listItem)
            End If
        Next

    End Sub

    ' Change the page to the selected display mode. 
    Public Sub ddlDisplayMode_SelectedIndexChanged(ByVal sender As Object, ByVal e As EventArgs)

        Dim selectedMode As String = ddlDisplayMode.SelectedValue
        Dim mode As WebPartDisplayMode = webPartManager.SupportedDisplayModes(selectedMode)

        If mode IsNot Nothing Then
            webPartManager.DisplayMode = mode
        End If

    End Sub

    ' Set the selected item equal to the current display mode. 
    Protected Sub Page_PreRender1(ByVal sender As Object, ByVal e As System.EventArgs) Handles Me.PreRender

        Dim items As ListItemCollection = ddlDisplayMode.Items
        Dim selectedIndex As Integer = items.IndexOf(items.FindByText(webPartManager.DisplayMode.Name))
        ddlDisplayMode.SelectedIndex = selectedIndex

    End Sub
End Class
